File: analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MethodDeclarationsAnalyzer/TestMethodDeclarations.cs
namespace Samples.CSharp;

using Microsoft.VisualStudio.TestTools.UnitTesting;

public class Address
{
    [TestMethod]
    public string GetZipCode() => "12345";
}

public struct Person
{
    [TestMethod]
    public string GetFullName() => "John Doe";
}

public record class Company
{
    [TestMethod]
    public string GetCompanyName() => "ACME";
}

public record struct Employee
{
    [TestMethod]
    public string GetEmployeeName() => "Jane Doe";
}

public enum EnumDeclaration
{
    Value1,
    Value2
}

public class Visibility
{
    [TestMethod]
    public void PublicMethod() { }

    [TestMethod]
    protected internal void ProtectedInternalMethod() { }

    [TestMethod]
    protected void ProtectedMethod() { }

    [TestMethod]
    internal void InternalMethod() { }

    [TestMethod]
    private protected void PrivateProtectedMethod() { }

    [TestMethod]
    private void PrivateMethod() { }

    [TestMethod]
    void NoAccessModifierMethod() { }

    internal class InternalClass
    {
        [TestMethod]
        public void Method() { }
    }

    private class PrivateClass
    {
        [TestMethod]
        public void Method() { }
    }
}

file class FileClass
{
    [TestMethod]
    public void Method() { }
}

class NoModifiers
{
    [TestMethod]
    public void Method() { }
}

public class MultipleMethods
{
    [TestMethod]
    public void Method1() { }

    [TestMethod]
    public void Method2() { }
}

public class Overloads
{
    [TestMethod]
    public void Method() { }

    [TestMethod]
    public void Method(int i) { }

    [TestMethod]
    public void Method(string s) { }

    [TestMethod]
    public void Method(int i, string s) { }
}

public class GenericClass<T>
{
    [TestMethod]
    public void Method() { }

    [TestMethod]
    public void Method<U>() { }
}

public class WithGenericMethod
{
    [TestMethod]
    public void Method<T>() { }
}

public partial class PartialClass : BaseClass
{
    [TestMethod]
    public void InFirstFile() { }

    public partial void PartialMethod(); // Test method attribute is defined in the other file.
}

public class PropertiesAndIndexers
{
    private int[] values;

    public string Property { get; set; }

    public int this[int i]
    {
        get => values[i];
        set => values[i] = value;
    }
}

public class LocalFunctions
{
    [TestMethod]
    public void Main()
    {
        [TestMethod]
        void LocalFunction() { }
    }
}

public abstract class BaseClass
{
    [TestMethod]
    public void BaseClassMethod() { }

    [TestMethod]
    public virtual void Method() { }
}

public class DerivedClass : BaseClass
{
    [TestMethod]
    public override void Method() { }
}

public class MultipleLevelInheritance : DerivedClass
{
    [TestMethod]
    public void MultipleLevelInheritanceMethod() { }
}

public interface IInterfaceWithTestDeclarations
{
    [TestMethod]
    public string GetZipCode();
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/Utilities/MethodDeclarationsAnalyzer/TestMethodDeclarations.vb
Imports Microsoft.VisualStudio.TestTools.UnitTesting

Namespace Samples.VB

    Public Class Address
        <TestMethod>
        Public Function GetZipCode() As String
            Return "12345"
        End Function
    End Class

    Public Structure Person
        <TestMethod>
        Public Function GetFullName() As String
            Return "John Doe"
        End Function
    End Structure

    Public Enum EnumDeclaration
        Value1
        Value2
    End Enum

    Public Class Visibility
        <TestMethod>
        Public Sub PublicMethod()
        End Sub

        <TestMethod>
        Protected Friend Sub ProtectedFriendMethod()
        End Sub

        <TestMethod>
        Protected Sub ProtectedMethod()
        End Sub

        <TestMethod>
        Friend Sub FriendMethod()
        End Sub

        <TestMethod>
        Private Protected Sub PrivateProtectedMethod()
        End Sub

        <TestMethod>
        Private Sub PrivateMethod()
        End Sub

        <TestMethod>
        Sub NoAccessModifierMethod()
        End Sub

        Friend Class FriendClass
            <TestMethod>
            Public Sub Method()
            End Sub
        End Class

        Private Class PrivateClass
            <TestMethod>
            Public Sub Method()
            End Sub
        End Class
    End Class

    Class NoModifiers
        <TestMethod>
        Public Sub Method()
        End Sub
    End Class

    Public Class MultipleMethods
        <TestMethod>
        Public Sub Method1()
        End Sub

        <TestMethod>
        Public Sub Method2()
        End Sub
    End Class

    Public Class OverloadedMethods
        <TestMethod>
        Public Sub Method()
        End Sub

        <TestMethod>
        Public Sub Method(i As Integer)
        End Sub

        <TestMethod>
        Public Sub Method(s As String)
        End Sub

        <TestMethod>
        Public Sub Method(i As Integer, s As String)
        End Sub
    End Class

    Public Class GenericClass(Of T)
        <TestMethod>
        Public Sub Method()
        End Sub

        <TestMethod>
        Public Sub Method(Of U)()
        End Sub
    End Class

    Public Class WithGenericMethod
        <TestMethod>
        Public Sub Method(Of T)()
        End Sub
    End Class

    Partial Public Class PartialClass
        Inherits BaseClass

        <TestMethod>
        Public Sub InFirstFile()
        End Sub

        Private Partial Sub PartialMethod() ' Test method attribute is defined in the other file.
        End Sub
    End Class

    Public Class PropertiesAndIndexers
        Private values() As Integer

        Public Property [Property] As String

        Default Public Property Item(index As Integer) As Integer
            Get
                Return values(index)
            End Get
            Set(value As Integer)
                values(index) = value
            End Set
        End Property
    End Class

    Public MustInherit Class BaseClass
        <TestMethod>
        Public Sub BaseClassMethod()
        End Sub

        <TestMethod>
        Public Overridable Sub Method()
        End Sub
    End Class

    Public Class DerivedClass
        Inherits BaseClass

        <TestMethod>
        Public Overrides Sub Method()
        End Sub
    End Class

    Public Class MultipleLevelInheritance
        Inherits DerivedClass

        <TestMethod>
        Public Sub MultipleLevelInheritanceMethod()
        End Sub
    End Class

    Public Interface IInterfaceWithTestDeclarations
        <TestMethod>
        Function GetZipCode() As String
    End Interface

End Namespace
